instrumentation/github.com/confluentinc/confluent-kafka-go/kafka/splunkkafka/producer.go

```diff
@@ -0,0 +1,148 @@
+// Copyright Splunk Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !windows
+// +build !windows
+
+package splunkkafka
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// A Producer wraps a kafka.Producer and traces its operations.
+type Producer struct {
+	*kafka.Producer
+	cfg            config
+	produceChannel chan *kafka.Message
+}
+
+// NewProducer calls kafka.NewProducer and wraps the resulting Producer with
+// tracing instrumentation.
+func NewProducer(conf *kafka.ConfigMap, opts ...Option) (*Producer, error) {
+	p, err := kafka.NewProducer(conf)
+	if err != nil {
+		return nil, err
+	}
+	return WrapProducer(p, opts...), nil
+}
+
+// WrapProducer wraps a kafka.Producer so that any produced events are traced.
+func WrapProducer(p *kafka.Producer, opts ...Option) *Producer {
+	wrapped := &Producer{
+		Producer: p,
+		cfg:      newConfig(opts...),
+	}
+	wrapped.produceChannel = wrapped.traceProduceChannel(p.ProduceChannel())
+	return wrapped
+}
+
+func (p *Producer) traceProduceChannel(out chan *kafka.Message) chan *kafka.Message {
+	if out == nil {
+		return nil
+	}
+
+	in := make(chan *kafka.Message, cap(out))
+	go func() {
```

Comment:
EOL of this goroutine is not synchronized with `Close` so users can miss the last `span.End()`.

I think something like this is safer (better synchronized):

```go
// in func (p *Producer) traceProduceChannel:
p.wg.Add(1) // wg is a new field in Producer to synchronize EOL of the goroutine scheduled below
go func() {
  for { 
    select {
      case <-p.closeSig:
        p.wg.Done()
        return
      case msg := <- in
        span := p.startSpan(msg)
        out <- msg
        span.End()
}()

// Close calls the wrapped Producer.Close and closes the producer channel.
func (p *Producer) Close() {
	close(p.produceChannel)
	p.Producer.Close()
	p.wg.Wait()
}
```

Reply:
moreover, thanks to this change the caller goroutine should panic if it tries ti send a message after closing the producer

Reply:
You can create an issue out of it if you do not want to spend time on this right now.